Restore most panel customizations

andrunko opened the pull request:
This adds most panel customizations to the panel extension, still missing some bits here and there (e.g. automatic updates indicator) but those will be added separately.

Depends on https://github.com/endlessm/gnome-shell/pull/685.

https://phabricator.endlessm.com/T30445

Commit:
layout: Move panel to bottom

https://phabricator.endlessm.com/T30445

Commit:
panel: Add infrastructure to allow updating the panel

https://phabricator.endlessm.com/T30445

Commit:
workspaceMonitor: Add helper wrapper to monitor apps/windows

https://phabricator.endlessm.com/T30445

Commit:
overview: Add helper wrapper to toggle between different overview views

https://phabricator.endlessm.com/T30445

Commit:
metadata: Align name/description with desktop extension

https://phabricator.endlessm.com/T30445

Commit:
endlessButton: Add endless button panel indicator

Original code imported from Endless 3.8 gnome-shell, with some
small changes to integrate with the extension (e.g. load icons
from extension dir).

https://phabricator.endlessm.com/T30445

Commit:
panelMenu: Add SingleIconMenuButton helper class

Original code imported from Endless 3.8 gnome-shell.

https://phabricator.endlessm.com/T30445

Commit:
hotCorner: Add hotcorner panel indicator

Original code imported from Endless 3.8 gnome-shell, with some
changes to integrate with the extension (e.g. load icons
from extension dir) and to remove the Enable/Disable menus
which had no effect.

https://phabricator.endlessm.com/T30445

Commit:
userMenu: Add user menu panel indicator

Original code imported from Endless 3.8 gnome-shell, with some
changes to:
- integrate with the extension (e.g. load icons from extension dir)
- allow launching Yelp using both org.gnome.Yelp.desktop (main) and
  yelp.desktop (alternative in case main app is not found - used on
  Fedora)
- hide the settings item from the child system indicator given we
  already have a separate menu for settings

https://phabricator.endlessm.com/T30445

Commit:
powerButton: Add power button panel indicator

Original code imported from Endless 3.8 gnome-shell, with some
small changes to rename from PowerMenu to PowerButton given this
is a single button.

https://phabricator.endlessm.com/T30445

Commit:
systemMenu: Add system menu panel indicator

Original code imported from Endless 3.8 gnome-shell and slightly
modified to remove automatic updates child indicator that will be
re-added later.

https://phabricator.endlessm.com/T30445

Commit:
power: Add specialized power/battery indicator and use it on system menu

The upstream indicator won't hide the indicator when the battery
is not present, let's specialize it to do so.

https://phabricator.endlessm.com/T30445

Commit:
systemMenu: Update volume indicator when session mode changes

In some cases when changing the session mode the volume indicator
gets hidden and it will require some event (e.g. volume change) to
get it displayed again.

This change makes sure the status is updated when the session mode
changes.

https://phabricator.endlessm.com/T30445

Commit:
appIconBar: Add app icon bar (taskbar) panel indicator

Original code imported from Endless 3.8 gnome-shell, with some
changes to:
- properly disconnect signals on destruction
- not use x_fill, no long supported
- remove the custom 'activate-icon-N' keybindings (for this one we
  may either re-add later or maybe just use existing upstream
  'switch-to-application-N' settings instead)

https://phabricator.endlessm.com/T30445

Commit:
stylesheet: Theme the world

Original theme imported from Endless 3.8 gnome-shell.

https://phabricator.endlessm.com/T30445

Commit:
panel: Customize the panel layout depending on session mode

Layout based on Endless 3.8 gnome-shell.

https://phabricator.endlessm.com/T30445

Commit:
overview: Add no windows dialog support

Original code imported from Endless 3.8 gnome-shell.

https://phabricator.endlessm.com/T30445

Commit:
appIconBar: Reload favorites' icons when favorites change

This slightly changes behaviour when an icon is pinned, which now
should immediately move the icon to the last position of the favorites,
before any running (and not pinned) app icon.

https://phabricator.endlessm.com/T30445

Commit:
appIconBar: Move icon to last position when unpinning

In line with the previous change, this also slightly changes
the behaviour by always moving unpinned items to the last
position (after other running apps) if the respective app is
running.

Given newly started apps are added to the last position,
let's do the same when unpinning.

https://phabricator.endlessm.com/T30445

Commit:
systemMenu: Update indicators style when session mode changes

https://phabricator.endlessm.com/T30445

Commit:
utils: Use a different key to store overrides

Using the same key name would clash with the Hack extension.
Switch to _panelFnOverrides instead.

https://phabricator.endlessm.com/T30445

andrunko commented:
Updated the PR with a new commit to change `Utils.override` to use a different key instead of  `_fnOverrides`.

Commit:
panel: Do not destroy our custom indicators when disabling the extension

This works because the panel keeps a ref to these indicators and
only deletes them when the panel itself is deleted (when the session is
reloaded), but it also means that these indicators instances are left in
memory even if the panel extension is disabled - note that the panel
extension won't recreate these indicators on disable/enable, given it
first checks if the instances are already available in the panel, which
would be the case when re-enabling this extension as we didn't destroy
them when disabling.

https://phabricator.endlessm.com/T30445
https://phabricator.endlessm.com/T30625

Commit:
panel: Hide our custom indicators when extension is disabled

We need to hide the indicators before restoring the original method
to make sure our custom indicators are hidden when the extension is
disabled now that we don't destroy the indicators when disabling anymore.

https://phabricator.endlessm.com/T30445
https://phabricator.endlessm.com/T30625

andrunko commented:
Force pushed to update one commit msg and also added 2 new commits to avoid destroying our indicators when disabling the extension - see https://phabricator.endlessm.com/T30625.

GeorgesStavracas commented on ui/layout.js:
```diff
@@ -0,0 +1,76 @@
+            this._rightPanelBarrier = null;
+        }
+
+        let monitor = this.primaryMonitor;
```
`const`

andrunko replied:
Done.

GeorgesStavracas commented on ui/layout.js:
```diff
@@ -0,0 +1,76 @@
+        const original = Utils.original(Layout.LayoutManager, '_updateBoxes');
+        original.bind(this)();
+
+        let monitor = this.primaryMonitor;
```
`const`

andrunko replied:
Done.

GeorgesStavracas commented on ui/panel.js:
```diff
@@ -0,0 +1,151 @@
+
+    let currentMode = sessionMode.currentMode;
+    let parentMode = sessionMode.parentMode;
+    let panelMode = _panelModes[currentMode] ?
```
All those 3 variables are `const`

andrunko replied:
Done.

GeorgesStavracas commented on ui/panel.js:
```diff
@@ -0,0 +1,151 @@
+
+function enable() {
+    Utils.override(Panel.Panel, '_updatePanel', function() {
+        let [panel, panelStyle] = _getPanelForSessionMode(Main.sessionMode);
```
`const`

andrunko replied:
Done.

Commit:
utils: Add helper method to load dbus interface from XML

https://phabricator.endlessm.com/T30445

Commit:
automaticUpdates: Add automatic updates panel indicator

Original code imported from Endless 3.8 gnome-shell, with some
changes to:
- integrate with the extension (e.g. load icons from extension dir)
- properly disconnect signals on destruction
- use utils.loadInterfaceXML to load dbus interfaces from extension dir
- some general cleanup

https://phabricator.endlessm.com/T30445

GeorgesStavracas commented on ui/panel.js:
```diff
@@ -0,0 +1,151 @@
+        return indicator;
+    });
+
+    let panel = Main.panel;
```
`const`

andrunko replied:
Done.

GeorgesStavracas commented on ui/panel.js:
```diff
@@ -0,0 +1,151 @@
+function disable() {
+    Utils.restore(Panel.Panel);
+
+    for (var i = _extraIndicators.length - 1; i >= 0; i--) {
```
I think this could be simplified with

```javascript
for (const indicator of _extraIndicators.reverse())
    indicator.destroy();
```

andrunko replied:
This code has changed in the latest version (we dont destroy indicators anymore, but we do clear the destroy signal sources we create) please check again.

GeorgesStavracas commented on ui/panel.js:
```diff
@@ -0,0 +1,151 @@
+    }
+    _extraIndicators = [];
+
+    if (_sessionModeUpdatedId) {
```
`if (_sessionModeUpdateId > 0) {` for extra clarity

andrunko replied:
Tbh I prefer to use `if (_sessionModeUpdatedId)` because it still works if we never set the variable - although not the case here this is the case in some indicators for example that only set these "source id" variables at a later point after construction (e.g. after a D-Bus proxy gets connected) and never set a default value for these variables on constructor.

That said I would prefer to leave it as is for consistency.

GeorgesStavracas commented on ui/panel.js:
```diff
@@ -0,0 +1,151 @@
+        this.statusArea[role] = indicator;
+        _extraIndicators.push(indicator);
+
+        let destroyId = indicator.connect('destroy', emitter => {
```
`const`

andrunko replied:
Done.

GeorgesStavracas commented on ui/panel.js:
```diff
@@ -0,0 +1,151 @@
+        _extraIndicators.push(indicator);
+
+        let destroyId = indicator.connect('destroy', emitter => {
+            let index = _extraIndicators.indexOf(indicator);
```
`const`

andrunko replied:
Done.

GeorgesStavracas commented on ui/panel.js:
```diff
@@ -0,0 +1,151 @@
+
+        original.bind(this)();
+
+        for (let role in EXTRA_PANEL_ITEM_IMPLEMENTATIONS) {
```
`const role`

andrunko replied:
Done.

GeorgesStavracas commented on ui/panel.js:
```diff
@@ -0,0 +1,151 @@
+// -*- mode: js; js-indent-level: 4; indent-tabs-mode: nil -*-
```
These headers aren't necessary anymore these days

andrunko replied:
I really would prefer to not remove these as they are simply in every js file - if you insist I'd add a new commit removing it from all files at once but I wouldn't bother - btw, why it it not needed anymore?

Commit:
systemMenu: Add automatic updates indicator

The layout/position is the same as in Endless 3.8 gnome-shell.

https://phabricator.endlessm.com/T30445

GeorgesStavracas commented on ui/endlessButton.js:
```diff
@@ -0,0 +1,149 @@
+
+        let iconFile = Gio.File.new_for_uri(
+            `file://${PanelExtension.path}/data/icons/endless-button-symbolic.svg`);
+        this.setIcon(new Gio.FileIcon({ file: iconFile }));
```
```javascript
const file = Gio.File.new_for_uri(
    `file://${PanelExtension.path}/data/icons/endless-button-symbolic.svg`);
this.setIcon(new Gio.FileIcon({ file }));
```

andrunko replied:
Done.

andrunko commented:
Added 3 new commits to include the automatic updates indicator.

GeorgesStavracas commented on ui/endlessButton.js:
```diff
@@ -0,0 +1,149 @@
+
+        let pageChangedId = Main.overview.connect('page-changed', this._onOverviewPageChanged.bind(this));
+        let showingId = Main.overview.connect('showing', this._onOverviewShowing.bind(this));
+        let hidingId = Main.overview.connect('hiding', this._onOverviewHiding.bind(this));
```
`const` for all three of them

andrunko replied:
Done.

GeorgesStavracas commented on ui/endlessButton.js:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    _updateHoverLabel(hiding) {
+        let viewSelector = Main.overview.viewSelector;
```
`const`

andrunko replied:
Done.

GeorgesStavracas commented on ui/endlessButton.js:
```diff
@@ -0,0 +1,149 @@
+
+            // Update the tooltip position
+            let monitor = Main.layoutManager.findMonitorForActor(this._label);
+            let iconMidpoint = this.get_transformed_position()[0] + this.width / 2;
```
`const` for both

andrunko replied:
Done.

GeorgesStavracas commented on ui/endlessButton.js:
```diff
@@ -0,0 +1,149 @@
+        if (!hiding &&
+            viewSelector &&
+            viewSelector.getActivePage() === ViewSelector.ViewPage.APPS)
+            showingDesktop = false;
```
This could probably be simplified as

```javascript
const showindDesktop =
    hiding ||
    !viewSelector ||
    viewSelector.getActivePage() !== ViewSelector.ViewPage.APPS;
```

andrunko replied:
Done.

GeorgesStavracas commented on ui/endlessButton.js:
```diff
@@ -0,0 +1,149 @@
+        this.connect('destroy', () => {
+            Main.overview.disconnect(pageChangedId);
+            Main.overview.disconnect(showingId);
+            Main.overview.disconnect(hidingId);
```
The label should probably be destroyed here too

andrunko replied:
Good catch, done.

GeorgesStavracas commented on ui/hotCorner.js:
```diff
@@ -0,0 +1,54 @@
+        super._init(_('Hot Corner'), Clutter.ActorAlign.END, Clutter.ActorAlign.END);
+        this.add_style_class_name('hot-corner');
+
+        let iconFile;
```
Use `let file` here...

andrunko replied:
Done.

GeorgesStavracas commented on ui/hotCorner.js:
```diff
@@ -0,0 +1,54 @@
+            iconFile = Gio.File.new_for_uri(`file://${PanelExtension.path}/data/icons/hot-corner-rtl-symbolic.svg`);
+        else
+            iconFile = Gio.File.new_for_uri(`file://${PanelExtension.path}/data/icons/hot-corner-symbolic.svg`);
+        this.setIcon(new Gio.FileIcon({ file: iconFile }));
```
And simplify this to `this.setIcon(new Gio.FileIcon({ file }))`

andrunko replied:
Done.

GeorgesStavracas commented on ui/hotCorner.js:
```diff
@@ -0,0 +1,54 @@
+    vfunc_event(event) {
+        if (event.type() === Clutter.EventType.TOUCH_BEGIN ||
+            event.type() === Clutter.EventType.BUTTON_PRESS) {
+            let button = event.get_button();
```
`const`

andrunko replied:
Done.

GeorgesStavracas commented on ui/userMenu.js:
```diff
@@ -0,0 +1,306 @@
+              Shell.AppSystem.get_default().lookup_app(HELP_CENTER_LAUNCHER_ALT);
+        if (app) {
+            iconFile = Gio.File.new_for_uri(
+                `file://${PanelExtension.path}/data/icons/endless-help-symbolic.svg`);
```
I'm not sure if gettext is able to properly handle translations when backticks are present in the source file, this needs double checking

andrunko replied:
I guess this comment came to the wrong line? There is no translatable string involved here but good general point anyway, will check for occurrences.

andrunko replied:
Checked and couldn't find any place where we use backticks for translatable strings, please let me know if I am missing something.

GeorgesStavracas commented on ui/userMenu.js:
```diff
@@ -0,0 +1,306 @@
+
+        this.accessible_role = Atk.Role.MENU;
+
+        let menuLayout = new Panel.AggregateLayout();
```
`const`

andrunko replied:
Done.

GeorgesStavracas commented on ui/userMenu.js:
```diff
@@ -0,0 +1,306 @@
+        this.menu.addMenuItem(this._accountSection);
+        this.menu.addMenuItem(new PopupMenu.PopupSeparatorMenuItem());
+
+        let menuItemsSection = new PopupMenu.PopupMenuSection();
```
`const`

andrunko replied:
Done.

GeorgesStavracas commented on ui/userMenu.js:
```diff
@@ -0,0 +1,306 @@
+        super();
+
+        // User account's icon
+        this.userIconItem = new PopupMenu.PopupBaseMenuItem({
```
It is a good practice to keep field members private (thus, `this._userIconItem`) and add public getters and setters

andrunko replied:
Done, just prefixed with "_" and updated the code using it (only in the same file so didn't bother adding getters/setters).

GeorgesStavracas commented on ui/userMenu.js:
```diff
@@ -0,0 +1,306 @@
+        });
+        this._avatar.x_align = Clutter.ActorAlign.CENTER;
+
+        let iconButton = new St.Button({ child: this._avatar });
```
`const`

andrunko replied:
Done.

GeorgesStavracas commented on ui/userMenu.js:
```diff
@@ -0,0 +1,306 @@
+        this.addMenuItem(this.userIconItem);
+
+        // User account's name
+        this.userLabelItem = new PopupMenu.PopupBaseMenuItem({
```
`this._userLabelItem`

andrunko replied:
Done.

GeorgesStavracas commented on ui/userMenu.js:
```diff
@@ -0,0 +1,306 @@
+        this.menu.close(BoxPointer.PopupAnimation.NONE);
+        Main.overview.hide();
+
+        let app = Shell.AppSystem.get_default().lookup_app(launcherName);
```
`const`

andrunko replied:
Done.

GeorgesStavracas commented on ui/powerButton.js:
```diff
@@ -0,0 +1,54 @@
+
+        this._session = new GnomeSession.SessionManager();
+
+        let icon = new Gio.ThemedIcon({ name: 'system-shutdown-symbolic' });
```
`const`

andrunko replied:
Done.

GeorgesStavracas commented on ui/systemMenu.js:
```diff
@@ -0,0 +1,97 @@
+            y_align: Clutter.ActorAlign.CENTER,
+        });
+
+        let userMode = Main.sessionMode.hasOverview;
```
`const`

andrunko replied:
Done.

GeorgesStavracas commented on ui/systemMenu.js:
```diff
@@ -0,0 +1,97 @@
+        super._init(0.0, C_("System menu", "System"), false);
+        this.menu.actor.add_style_class_name('aggregate-menu');
+
+        let menuLayout = new Panel.AggregateLayout();
```
`const`

andrunko replied:
Done.

GeorgesStavracas reviewed (changes requested):
I'm far from the finish line, and the AppIconBar part is going to be specially tricky to review, but this is what I have for today. Will continue next week.

Commit:
appIconBar: Use const instead of let where appropriate

https://phabricator.endlessm.com/T30445

Commit:
automaticUpdates: Use const instead of let where appropriate

https://phabricator.endlessm.com/T30445

Commit:
endlessButton: Use const instead of let where appropriate

https://phabricator.endlessm.com/T30445

Commit:
hotCorner: Use const instead of let where appropriate

https://phabricator.endlessm.com/T30445

Commit:
layout: Use const instead of let where appropriate

https://phabricator.endlessm.com/T30445

Commit:
overview: Use const instead of let where appropriate

https://phabricator.endlessm.com/T30445

Commit:
panel: Use const instead of let where appropriate

https://phabricator.endlessm.com/T30445

Commit:
powerButton: Use const instead of let where appropriate

https://phabricator.endlessm.com/T30445

Commit:
systemMenu: Use const instead of let where appropriate

https://phabricator.endlessm.com/T30445

Commit:
userMenu: Use const instead of let where appropriate

https://phabricator.endlessm.com/T30445

Commit:
endlessButton: Also destroy tooltip label when button is destroyed

https://phabricator.endlessm.com/T30445

Commit:
userMenu: Prefix private variables with _

https://phabricator.endlessm.com/T30445

andrunko commented:
Thanks for the initial review, fixed most of the comments above and replied on the remaining ones for clarification. 

While on that also checked/fixed all other places I could find that should be using `const` instead of `let`.

In general I'd say we should try avoiding nitpicking code imported from the old (3.8) shell (specially the icon bar which is the bigger one), otherwise we would be reviewing code already reviewed... although of course if you see anything actually wrong we should fix it.

Commit:
appIconBar: Properly show running (and focused) app indicator

For some reason setting 'y_align' on the container is making it
ignore 'y_expand' (set to true) which in turn makes the
container too small to show the running app indicator (set via CSS
to a fixed position that considers the container will have the
same height as the panel).

Simply not setting 'y_align' makes things work as expected.

https://phabricator.endlessm.com/T30625

andrunko commented:
Pushed a new commit to fix an issue showing the running/focused app indicator in the app icon bar.

GeorgesStavracas reviewed (approved):
Most of the review comments that were marked as done are not actually done. In any case, you seem to be in a hurry to land this, and I'm not going to be the roadblock for this.

andrunko commented:
> Most of the review comments that were marked as done are not actually done. In any case, you seem to be in a hurry to land this, and I'm not going to be the roadblock for this.

Could you please point to the ones that were not done? Despite the hurry I wasnt planning to or didnt miss any review comment on purpose, I actually changed more places where I found related issues like const vs let, so please do let me know and I will work on the fixes.

I just double checked and could not find the ones that I said it was done that was not actually the case, curious about what I missed.
